packages/cli/src/cmds/preview/npmInstallHandler.ts

```diff
@@ -0,0 +1,120 @@
+      const tasks: Map<string, Promise<number | null>> = new Map<string, Promise<number | null>>();
+      if (ProjectSettingsHelper.isSpfx(projectSettings)) {
+        tasks.set("spfx", runTask(TaskDefinition.spfxInstall(workspaceFolder)));
+      } else {
+        if (!excludeFrontend && ProjectSettingsHelper.includeFrontend(projectSettings)) {
+          tasks.set("frontend", runTask(TaskDefinition.frontendInstall(workspaceFolder)));
+        }
+        if (!excludeBackend && ProjectSettingsHelper.includeBackend(projectSettings)) {
+          tasks.set("backend", runTask(TaskDefinition.backendInstall(workspaceFolder)));
+        }
+        if (!excludeBot && ProjectSettingsHelper.includeBot(projectSettings)) {
+          tasks.set("bot", runTask(TaskDefinition.botInstall(workspaceFolder)));
+        }
+      }
+
+      if (tasks.size > 0) {
+        cliLogger.necessaryLog(LogLevel.Info, constants.automaticNpmInstallHintMessage);
+        try {
+          const properties: { [key: string]: string } = {};
+          for (const key of tasks.keys()) {
+            properties[key] = "true";
+          }
+          cliTelemetry.sendTelemetryEvent(TelemetryEvent.AutomaticNpmInstallStart, properties);
+        } catch {
+          // ignore telemetry error
+        }
+
+        const properties: { [key: string]: string } = {};
+        let failed = false;
+        const keys = tasks.keys();
+        for (const task of tasks.values()) {
+          const code = await task;
+          if (code !== 0) {
+            failed = true;
+          }
+          properties[keys.next().value] = code + "";
+        }
+        if (failed) {
+          cliTelemetry.sendTelemetryErrorEvent(
+            TelemetryEvent.AutomaticNpmInstall,
+            NpmInstallFailed(),
+            properties
+          );
+        } else {
+          properties[TelemetryProperty.Success] = TelemetrySuccess.Yes;
+          cliTelemetry.sendTelemetryEvent(TelemetryEvent.AutomaticNpmInstall, properties);
+        }
+      }
+    }
+  } catch (error) {
+    cliLogger.warning(`Automatic npm install failed: ${error}`);
+  }
+}
+
+export function getAutomaticNpmInstallSetting(): boolean {
+  try {
+    const result = UserSettings.getConfigSync();
+    if (result.isErr()) {
+      throw result.error;
+    }
+
+    const config = result.value;
+    const automaticNpmInstallOption = "automatic-npm-install"; // TODO: use CliConfigOptions.AutomaticNpmInstall instead
+    if (!(automaticNpmInstallOption in config)) {
+      return false; // TODO: make automatic-npm-install enabled by default
+    }
+    return config[automaticNpmInstallOption] !== CliConfigAutomaticNpmInstall.Off;
+  } catch (error: any) {
+    cliLogger.warning(`Getting automatic-npm-install setting failed: ${error}`);
+    return false; // TODO: make automatic-npm-install enabled by default
+  }
+}
+
+async function runTask(task: ITaskDefinition): Promise<number | null> {
```

Comment:
What's this different from `preview/task.ts`? Or, how about just reuse `preview/task.ts task.wait()`

Reply:
`preview/task.ts` is not that extensible to reuse... it hides stdout, which is required by auto npm install command to print npm logs. Considered to refactor `preview/task.ts` to support this case afterwards.

Reply:
I see, make sense.